File: src/utils/secureStoreChunks.ts
export const SECURE_STORE_CHUNK_SIZE = 1800;
export const SECURE_STORE_MAX_CHUNKS = 20;
export const SECURE_STORE_TOO_LARGE_ERROR = 'Supabase auth session is too large for SecureStore';

export type ChunkMetadata = {
  chunked: true;
  count: number;
};

export function chunkKey(key: string, index: number) {
  return `${key}.${index}`;
}

export function parseChunkMetadata(value: string | null): ChunkMetadata | null {
  if (!value) return null;
  try {
    const parsed = JSON.parse(value);
    if (
      parsed?.chunked === true &&
      Number.isInteger(parsed.count) &&
      parsed.count > 0 &&
      parsed.count <= SECURE_STORE_MAX_CHUNKS
    ) {
      return parsed as ChunkMetadata;
    }
  } catch {
    return null;
  }
  return null;
}

export function splitSecureStoreValue(value: string): string[] {
  const chunks = value.match(new RegExp(`.{1,${SECURE_STORE_CHUNK_SIZE}}`, 'g')) ?? [];
  if (chunks.length > SECURE_STORE_MAX_CHUNKS) {
    throw new Error(SECURE_STORE_TOO_LARGE_ERROR);
  }
  return chunks;
}

File: apps/android-native/app/src/main/java/com/wxxtae/pws/nativepreview/domain/SecureStoreChunks.kt
package com.wxxtae.pws.nativepreview.domain

const val SecureStoreChunkSize: Int = 1800
const val SecureStoreMaxChunks: Int = 20
const val SecureStoreTooLargeError: String = "Supabase auth session is too large for SecureStore"

data class ChunkMetadata(
    val chunked: Boolean,
    val count: Int,
)

object SecureStoreChunks {
    fun chunkKey(key: String, index: Int): String = "$key.$index"

    fun parseChunkMetadata(value: String?): ChunkMetadata? {
        if (value.isNullOrBlank()) return null
        val compact = value.replace(" ", "")
        if (!compact.contains("\"chunked\":true")) return null
        val countMatch = Regex("\"count\":(\\d+)").find(compact) ?: return null
        val count = countMatch.groupValues[1].toIntOrNull() ?: return null
        if (count <= 0 || count > SecureStoreMaxChunks) return null
        return ChunkMetadata(chunked = true, count = count)
    }

    fun splitSecureStoreValue(value: String): List<String> {
        if (value.isEmpty()) return emptyList()
        val chunks = value.chunked(SecureStoreChunkSize)
        if (chunks.size > SecureStoreMaxChunks) throw IllegalStateException(SecureStoreTooLargeError)
        return chunks
    }
}
